### PR TITLE
Input Plugins Framework and HTC Vive Support

### PR DESCRIPTION
**Pull request by hyperlogic**
**NOTE:** This is merge request into the plugins branch **NOT** master.

Added new **Avatar > Input Plugins** menu which adds the ability to enable and disable input device plugins.  It generally mirrors the Display Plugins architecture and the source is located in the libraries/input-plugins folder.  These are the current devices that use this plug-in system.

   * SDL2 for joysticks and joypads.
   * Keyboard and Mouse
   * Sixense Razor Hydra
   * HTC Vive Lighthouse Controllers

Also, there is a new ** View > Standing HMD Sensor Mode ** menu option.  Checking this option will enable "room scale" VR handling.  In this mode, your avatar motion will follow your head and you move in the world.  When disabled, the current behavior of leaning your avatar is maintained.  This mode *should* be HMD agnostic.  i.e. it works in the Rift as well as the Vive.

Known issues

   * The hemispherical overlay is difficult or impossible to use when in "Standing HMD Mode" using the HTC Vive.  Because it is located in the center of the "room" not around your head.  We are working on a fix.
   * When tracking is lost, while in "Standing HMD Mode" while using the HTC Vive, large teleportations can occur.

### CODE REVIEW

File: interface/src/Menu.cpp [hunk 1]
```diff
@@ -28,7 +28,7 @@
 #include "devices/DdeFaceTracker.h"
 #include "devices/Faceshift.h"
 #include "devices/RealSense.h"
-#include "devices/SixenseManager.h"
+#include <input-plugins\SixenseManager.h> // TODO: should be able to remove this once input plugin architecture is finished
```
Comment (jherico) [inline interface/src/Menu.cpp]: forward slash

File: libraries/shared/src/GLMHelpers.h [hunk 2]
```diff
@@ -150,4 +150,9 @@ T toNormalizedDeviceScale(const T& value, const T& size) {
 #define PITCH(euler) euler.x
 #define ROLL(euler) euler.z
 
+glm::mat4 createMatFromQuatAndPos(glm::quat q, glm::vec3 p);
```
Comment (jherico) [inline libraries/shared/src/GLMHelpers.h]: should be `(const glm::quat& q, const glm::vec3& p)`, same for the following functions.

File: libraries/input-plugins/src/input-plugins/ViveControllerManager.cpp [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+#include "GeometryCache.h"
+#include <gpu/Batch.h>
+#include <gpu/Context.h>
+#include <DeferredLightingEffect.h>
+#include <display-plugins\openvr\OpenVrHelpers.h>
+#include "NumericalConstants.h"
+#include "UserActivityLogger.h"
+
+extern vr::IVRSystem *_hmd;
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/ViveControllerManager.cpp]: conding standard is `extern vr::IVRSystem* _hmd;` as opposed to `extern vr::IVRSystem *_hmd;`.

File: libraries/input-plugins/src/input-plugins/SDL2Manager.cpp [hunk 4]
```diff
@@ -111,19 +120,21 @@ void SDL2Manager::update() {
...
-                    : HFActionEvent::endType();
-                    
-                    // global action events fire in the center of the screen
-                    Application* app = Application::getInstance();
-                    PickRay pickRay = app->getCamera()->computePickRay(app->getTrueMouseX(),
-                                                                       app->getTrueMouseY());
-                    HFActionEvent actionEvent(actionType, pickRay);
-                    qApp->sendEvent(qApp, &actionEvent);
                 }
+                // TODO: This will probably end up being deleted
```
Comment (jherico) [inline libraries/input-plugins/src/input-plugins/SDL2Manager.cpp]: delete dead code

**Comment by hifi-gustavo** [comment]
Build console output for commit: d985d0293a89efd2cebbcefb7da1d10e401fbb21 available at: https://gist.github.com/3b05979afdfc97da2e16<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5373.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5373.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a4cbb7f41258fdb90508f791b23e5ebe858c1832 available at: https://gist.github.com/3b05979afdfc97da2e16<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5373.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5373.dmg

**Pull request merged**